crates/biome_js_analyze/src/lint/nursery/no_nested_component_definitions.rs

```diff
@@ -0,0 +1,279 @@
+};
+use biome_console::markup;
+use biome_js_syntax::{
+    AnyJsAssignment, AnyJsAssignmentPattern, AnyJsBinding, AnyJsFunction, JsAssignmentExpression,
+    JsCallArgumentList, JsCallArguments, JsCallExpression, JsInitializerClause, JsSyntaxToken,
+    JsVariableDeclarator,
+};
+use biome_rowan::{AstNode, declare_node_union};
+use biome_string_case::Case;
+
+declare_lint_rule! {
+    /// Disallows defining React components inside other components.
+    ///
+    /// Component definitions inside other components cause them to be recreated on every render,
+    /// which can lead to performance issues and unexpected behavior.
+    ///
+    /// When a component is defined inside another component:
+    /// - It gets recreated on every render of the parent component
+    /// - It loses its internal state when the parent rerenders
+    /// - It defeats props memoization and optimization techniques
+    /// - It creates new function references on every render
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// A new component is created every time ParentComponent renders:
+    /// ```jsx,expect_diagnostic
+    /// function ParentComponent() {
+    ///   function ChildComponent() {
+    ///     return <div>Hello</div>;
+    ///   }
+    ///
+    ///   return <ChildComponent />;
+    /// }
+    /// ```
+    ///
+    /// Even with memo, a new component is still created on each render:
+    /// ```jsx,expect_diagnostic
+    /// function ParentComponent() {
+    ///   const MemoizedChild = memo(() => {
+    ///     return <div>Hello</div>;
+    ///   });
+    ///
+    ///   return <MemoizedChild />;
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// Component is defined outside other components:
+    /// ```jsx
+    /// function ChildComponent() {
+    ///   return <div>Hello</div>;
+    /// }
+    ///
+    /// function ParentComponent() {
+    ///   return <ChildComponent />;
+    /// }
+    /// ```
+    ///
+    /// ## Correct approaches
+    ///
+    /// 1. Move the component definition outside:
+    ///    ```jsx
+    ///    function ChildComponent() {
+    ///      return <div>Hello</div>;
+    ///    }
+    ///
+    ///    function ParentComponent() {
+    ///      return <ChildComponent />;
+    ///    }
+    ///    ```
+    ///
+    /// 2. Pass components as props:
+    ///    ```jsx
+    ///    function ParentComponent({ CustomComponent }) {
+    ///      return <CustomComponent />;
+    ///    }
+    ///    ```
+    ///
+    /// 3. Use React's Children API:
+    ///    ```jsx
+    ///    function ParentComponent({ children }) {
+    ///      return <div>{children}</div>;
+    ///    }
+    ///    ```
+    pub NoNestedComponentDefinitions {
+        version: "next",
+        name: "noNestedComponentDefinitions",
+        language: "js",
+        sources: &[RuleSource::EslintReact("no-nested-components")],
+        recommended: false,
+    }
```

Comment:
Make sure to put this in the react domain

Reply:
Thanks, done!